libp2p/src/builder/phase/tcp.rs

```diff
@@ -0,0 +1,208 @@
+use super::*;
+use crate::SwarmBuilder;
+#[cfg(all(
+    not(target_arch = "wasm32"),
+    any(feature = "tcp", feature = "websocket")
+))]
+use libp2p_core::muxing::{StreamMuxer, StreamMuxerBox};
+#[cfg(all(feature = "websocket", not(target_arch = "wasm32")))]
+use libp2p_core::Transport;
+#[cfg(all(
+    not(target_arch = "wasm32"),
+    any(feature = "tcp", feature = "websocket")
+))]
+use libp2p_core::{InboundUpgrade, Negotiated, OutboundUpgrade, UpgradeInfo};
+use std::marker::PhantomData;
+
+pub struct TcpPhase {}
+
+macro_rules! impl_tcp_builder {
+    ($providerKebabCase:literal, $providerPascalCase:ty, $path:ident) => {
+        #[cfg(all(
+            not(target_arch = "wasm32"),
+            feature = "tcp",
+            feature = $providerKebabCase,
+        ))]
+        impl SwarmBuilder<$providerPascalCase, TcpPhase> {
+            pub fn with_tcp<SecUpgrade, SecStream, SecError, MuxUpgrade, MuxStream, MuxError>(
```

Comment:
I'd like us to have some basic docs on here that explain to the user that `security_upgrade` and `multiplexer_upgrade` are function-pointers!